### PR TITLE
CLI testing, squashed, some minor improvements

### PR DESCRIPTION
**Pull request by eode**
This is basically pull request #193, squashed, and with recent changes merged.  It also has a minor bugfix, and has an env-var option "QUILT_TEST_CLI_SUBPROC=True", which, if set, will cause it to use subprocesses for testing, instead of the current (faster) method.

### CODE REVIEW

**Comment by asah** [comment]
@dimaryaz after recent breakage, I asked @eode to create a framework for testing the CLI, incl catching if the CLI accidentally changes.  From here, we can add detailed CLI tests for subtle details, e.g. error messages, return codes, etc.  This is a starter projectthis and his first commit.

@eode Dima is lead eng and can get backed up with code reviews.  I'll send you some more projects to stay incl bug fixes.  Thx!

File: quilt/tools/main.py [hunk 1]
```diff
@@ -139,7 +142,16 @@ def check_hash(group, hashstr):
...
+    return parser
+
+
+def main(args=None):
+    """Build and run parser
+
+    :param args: cli args from tests
+    """
+    parser = argument_parser()
+    args = parser.parse_args(args) if args is not None else parser.parse_args()
```
Comment (dimaryaz) [inline quilt/tools/main.py]: I think `args = parser.parse_args(args)` should be ok - `None` is already the default value, so it just works.

File: quilt/test/test_cli.py [hunk 2]
```diff
@@ -0,0 +1,652 @@
...
+                new[k] = _mcopy(v) if v else None
+            return new
+        return _mcopy(self)
+
+    def __repr__(self):
+        args = '\n'.join(repr(x) for x in self)
+        return 'ArgparseIntrospector with args:\n' + args
+
+
+class RecursiveMappingWrapper(collections.MutableMapping):
```
Comment (dimaryaz) [inline quilt/test/test_cli.py]: This feels really over-engineered to me, to be honest...

File: quilt/test/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+
+# inspect.argspec is deprecated, so
+try:
+    from funcsigs import signature  # python 2.7
+except ImportError:
+    from inspect import signature
+
+
+## "Static" vars
+_TEST_DIR = os.path.abspath(inspect.stack()[0][1])
```
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: ..because __file__ is unreliable in some versions of python.  I can just use __file__, though, if you'd rather.

File: quilt/test/test_cli.py [hunk 4]
```diff
@@ -0,0 +1,652 @@
...
+PACKAGE_DIR = os.path.join('/', *_TEST_DIR.split(os.path.sep)[:-3])
```
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: Yep, I thought I'd fixed that.  Thanks for the catch.

File: quilt/tools/main.py (hunk 1 above)
Comment (eode) [reply quilt/tools/main.py]: Not in the case where blank args are sent, if anyone ever tests what happens when `quilt` is called without args.  ..although, we may never test that.

File: quilt/tools/command.py [hunk 5]
```diff
@@ -67,8 +68,19 @@
...
-
+try:
+    VERSION = pkg_resources.require('quilt')[0].version
+except pkg_resources.DistributionNotFound:
+    # Not installed, so we'll get the version manually.
+    try:
+        mydir = os.path.abspath(stack()[0][1])
+        pkgdir = os.path.join('/', *mydir.split(os.path.sep)[:-3])
+        text = open(os.path.join(pkgdir, 'setup.py')).read()
+        text = text.split('\nQUILT_VERSION = ', 1)[1]
```
[the remark this answers is not in the document]
Comment (eode) [reply quilt/tools/command.py]: It is super-hacky, but at the time it was written, the goal was to enable invoking main.py from the tests directory so that pytest runs main.py, without an install, and without creating any new files -- although, some of those restrictions may have been made in part simply as questions on the job description.  Also, this was before I was really familiar with the codebase, and before I had a deeper understanding of what's needed.

Calling from the CLI is technically more complete, but it's significantly slower, and with modifications I've added to main.py, isn't absolutely necessary.  The likelihood that there will be a significant difference between doing a syscall of `['quilt'] + quilt_args` vs calling `main.argument_parser()` and `main.main(args)` is fairly small.

I'd be happy to either store the version in a separate version file and read that, or simply forego actual CLI calls altogether in favor of simulated CLI calls.

File: quilt/test/test_cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: Wait, no, I remember now -- I was thinking of using os.path.join(), which I did do -- os.path.split only splits off the last item.  I thought this was cleaner than `os.path.split(os.path.split(os.path.split(_TEST_DIR)[0])[0])[0]`

File: quilt/test/test_cli.py [hunk 6]
```diff
@@ -0,0 +1,652 @@
...
+    """Shows CLI param test coverage based on reporting by test functions.
+
+    Returns a dict with `'missing'` and `'percentage'` keys.
+    """
+    param_tree = RecursiveMappingWrapper(get_current_param_tree())
+    ideal_data = [tuple(v) for v in param_tree.iterpaths()]
+    ideal = set(ideal_data)
+    tested = set(tuple(v) for v in TESTED_PARAMS)
+    # remove covered test cases that aren't in the main argparser
+    tested = tested - tested.difference(ideal)
```
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: Yes, that would more ideally be an &.  >.>

**Comment by quiltdata** [comment]
Please add a *short* one line comment explaining the performance issue.  If
you can make it less hacky, that would great ideally moving the hacky code
from command.py to some file in test/ where hacks are (more) allowed.

File: quilt/test/test_cli.py (hunk 2 above)
Comment (eode) [reply quilt/test/test_cli.py]: It was a piece of code from my personal library that turns any mapping into a recursively-accessible mapping.  Since it was more than is needed, and we have no need for it elsewhere, I've taken just the snippets we need out, and integrated them into ArgparseIntrospector now.

File: quilt/test/test_cli.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: Fixed.

File: quilt/test/test_cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: Fixed.

File: quilt/test/test_cli.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (eode) [reply quilt/test/test_cli.py]: Fixed.

File: quilt/tools/command.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eode) [reply quilt/tools/command.py]: command.py and setup.py now read from a new VERSION file.  This seemed the least-hacky way of enabling running quilt when not installed.  If this isn't preferred, we can just remove the ability to run quilt when not installed.

File: quilt/tools/main.py (hunk 1 above)
Comment (eode) [reply quilt/tools/main.py]: I misread what you said there.  Fixed.

**Comment by eode** [comment]
@quiltdata Re: One-line comment about quilt performance issue:  That's issue #194, and I've circumvented it by defaulting to not using an actual CLI call, but using main.argument_parser() and main.main(args) instead.  Commented as #NOTE: in test_cli.py.

**Comment by eode** [comment]
@dimaryaz I think this is ready for a re-review.

File: compiler/quilt/tools/command.py [hunk 7]
```diff
@@ -68,7 +69,18 @@
 
 LOG_TIMEOUT = 3  # 3 seconds
 
-VERSION = pkg_resources.require('quilt')[0].version
+try:
+    VERSION = pkg_resources.require('quilt')[0].version
+except pkg_resources.DistributionNotFound:
+    # Not installed, so we'll get the version manually.
```
Comment (dimaryaz) [inline compiler/quilt/tools/command.py]: So I still don't understand: why is this needed? When would quilt not be installed?

File: compiler/setup.py [hunk 8]
```diff
@@ -12,9 +14,15 @@ def readme():
     """
     return readme_short
 
+
+def quilt_version():
+    pkgdir = os.path.split(os.path.abspath(__file__))[0]
```
Comment (dimaryaz) [inline compiler/setup.py]: `pkgdir = os.path.dirname(__file__)`

I don't think you even need absolute path; relative ones work, too.

File: compiler/setup.py [hunk 9]
```diff
@@ -12,9 +14,15 @@ def readme():
...
+    return open(os.path.join(pkgdir, 'VERSION')).read().strip()
```
Comment (dimaryaz) [inline compiler/setup.py]: Need to use `with open(...):`, so the fd is not leaked.

File: compiler/quilt/tools/command.py [hunk 10]
```diff
@@ -68,7 +69,18 @@
...
+    try:
+        filedir = os.path.split(os.path.abspath(__file__))[0]
+        quiltdir = os.path.split(filedir)[0]
+        pkgdir = os.path.split(quiltdir)[0]
+        VERSION = open(os.path.join(pkgdir, 'VERSION')).read().strip()
```
Comment (dimaryaz) [inline compiler/quilt/tools/command.py]: This looks similar to the helper function below?

File: compiler/quilt/test/test_cli.py [hunk 11]
```diff
@@ -0,0 +1,589 @@
...
+
+# inspect.argspec is deprecated, so
+try:
+    from funcsigs import signature  # python 2.7
+except ImportError:
+    from inspect import signature
+
+
+## "Static" vars
+_TEST_DIR = os.path.split(os.path.abspath(__file__))[0]
```
Comment (dimaryaz) [inline compiler/quilt/test/test_cli.py]: Use `os.path.dirname`

Comment (eode) [reply compiler/quilt/test/test_cli.py]: Original code:

@dimaryaz wrote:
> There's os.path.split (which also supports both / and \ on win32)

Was told by you to use os.path.split, over splitting on os.path.sep (which is /, or \ on win32).
Changed to your current preference, and changed other occurrences as well.

File: compiler/quilt/tools/command.py (hunk 7 above)
Comment (eode) [reply compiler/quilt/tools/command.py]: @eode wrote:
> It is super-hacky, but at the time it was written, the goal was to enable invoking main.py from the tests directory so that pytest runs main.py, without an install, and without creating any new files -- although, some of those restrictions may have been made in part simply as questions on the job description.

To be more clear:  I was instructed to do so.  It was one of the requirements I was initially given on this specific project.  

However, it may be a requirement that has has since changed, like the requirement on not creating any new files has (a `VERSION` file was permitted).

..in either case, it was what I was instructed to do, and I found a way to do it with the specified restrictions.  If you want that capacity removed, or if you'd like for me to use a different implementation, I'll follow the new requirements you set forth.

As to when quilt wouldn't be installed:  I think it was requested of me in order to fulfil a specific (or potential) testing or CI setup need, but I don't know the reason definitively.

File: compiler/quilt/tools/command.py (hunk 10 above)
Comment (eode) [reply compiler/quilt/tools/command.py]: * they both operate when not installed
* `quilt.tools.command` can't reasonably be imported from `setup.py`
* `setup.py` can't be reasonably imported from `command.py`

...can you be more specific with what you'd like me to do here?

File: compiler/setup.py (hunk 8 above)
Comment (eode) [reply compiler/setup.py]: Nope, I it's not necessary in this case, just done habitually.  Working with absolute paths isn't a bad habit, though, and it doesn't really obscure what's happening here.  Changed to your preference, anyway, as there's no need for it here.

File: compiler/setup.py (hunk 9 above)
Comment (eode) [reply compiler/setup.py]: In what case would the FD be leaked here?

* There's no potential for a circular reference here
* The file is opened and read immediately, then closes automatically when not assigned to a variable.
* The closest possible thing to a leak would be an exception in read(), which would result in a program exit
  * Even if that were the case, the FD would only remain open (held in the stack frame associated with the traceback at sys.last_traceback) until the next exception occurs.
    * Even if another exception didn't occur (unlikely) the FD would only remain open until setup exited, and would have no ill effects.
  * Even if the OS is hard-killing the process (KILL as opposed to INT on POSIX systems), then *all* open files are generally 'leaked', and it's up to the OS to reclaim those -- in any case, that's Python and the OS's job, not ours.

File: compiler/quilt/tools/command.py (hunk 7 above)
Comment (asah) [reply compiler/quilt/tools/command.py]: This was me -- at the time, I was cautious about requiring pip install to be run before unittests can pass... but more & more, I'm seeing that pip install is kinda required for even basic development.

So if you want to require this, I'm ok with it.  @dimaryaz gets the final say, of course.

Comment (eode) [reply compiler/quilt/tools/command.py]: ..sounds good.

Comment (dimaryaz) [reply compiler/quilt/tools/command.py]: Yeah, let's require quilt to be installed. We already need all of the requirements to be installed - so we're not saving anything by not requiring quilt itself.

Comment (eode) [reply compiler/quilt/tools/command.py]: Sounds good, I'll excise that code.

**Comment by eode** [comment]
..improved some commnets [sic].  Also, minor bugfix re: reading version using cli during Subprocess mode in test_cli.  Verified that this works with QUILT_TEST_CLI_SUBPROC both True and False.  This should be set to True on CI testing for more thorough testing of the cli.

File: compiler/quilt/tools/command.py [hunk 12]
```diff
@@ -11,6 +11,7 @@
 import json
 import os
 import re
+from inspect import stack
```
Comment (dimaryaz) [inline compiler/quilt/tools/command.py]: Delete this?

File: compiler/setup.py [hunk 13]
```diff
@@ -1,4 +1,6 @@
 from setuptools import setup, find_packages
+import os
```
Comment (dimaryaz) [inline compiler/setup.py]: Delete this

File: compiler/quilt/tools/main.py [hunk 14]
```diff
@@ -13,18 +13,23 @@
 from . import command
 from .const import DEFAULT_QUILT_YML
 
+# Importing this monkey-patches `command` if testing the CLI in subprocess mode
+from ..test import test_cli     # pylint:disable=W0611
```
Comment (dimaryaz) [inline compiler/quilt/tools/main.py]: Importing this in production code feels wrong... I think doing a conditional import and patching would be better. Also, make monkey-patching a helper function instead of a side-effect of import. So something like this:

```
if os.environ.get('QUILT_TEST_CLI_SUBPROC') == "True":
  from ..test import test_cli
  test_cli.monkey_patch(...)
```

Comment (eode) [reply compiler/quilt/tools/main.py]: Yeah, I originally did something similar.  I'll go back to what I had.  I was mainly just trying to get the explicitly testing-related code out of main.py.
